dali/operators/segmentation/random_object_bbox.cc

```diff
@@ -0,0 +1,461 @@
+      }
+      break;
+    default:
+      assert(!"Unreachable code");
+  }
+}
+
+template <typename Box>
+void StoreBox(const OutListCPU<int, 1> &out1,
+              const OutListCPU<int, 1> &out2,
+              RandomObjectBBox::OutputFormat format,
+              int sample_idx, Box &&box) {
+  StoreBox(out1, out2, format, sample_idx, box.lo, box.hi);
+}
+
+void RandomObjectBBox::GetBgFgAndWeights(
+      ClassVec &classes, WeightVec &weights, int &background, int sample_idx) {
+  background = background_[sample_idx].data[0];
+  if (ignore_class_)
+    return;  // we don't care about classes at all
+
+  if (classes_.IsDefined()) {
+    const auto &cls_tv = classes_[sample_idx];
+    int ncls = cls_tv.shape[0];
+    classes.resize(ncls);
+    if (!weights_.IsDefined()) {
+      weights.clear();
+      weights.resize(ncls, 1.0f);
+    }
+
+    int min_class_label = ncls ? cls_tv.data[0] : 0;
+    bool choose_different_background = false;
+    for (int i = 0; i < ncls; i++) {
+      classes[i] = cls_tv.data[i];
+      if (classes[i] < min_class_label)
+        min_class_label = classes[i];
+
+      if (classes[i] == background) {
+        if (background_.IsDefined()) {
+          // Background was explicitly specified this way - that's an error
+          DALI_FAIL(make_string("Class label ", classes[i],
+            " coincides with background label - please specify a different background label or "
+            " remove it from your list of foreground classes."));
+        } else {
+          // Not specified? Pick a different one.
+          choose_different_background = true;
+        }
+      }
+    }
+    if (choose_different_background) {
+      if (min_class_label == std::numeric_limits<int>::min()) {
+        background = std::numeric_limits<int>::max();  // wrap around
+
+        std::unordered_set<int> cls_set;
+        for (auto cls : classes)
+          cls_set.insert(cls);
+
+        while (cls_set.count(background))
+          background--;
+      } else {
+        background = min_class_label - 1;
+      }
+    }
+  }
+  if (weights_.IsDefined()) {
+    const auto &cls_tv = weights_[sample_idx];
+    int ncls = cls_tv.shape[0];
+    weights.resize(ncls);
+    for (int i = 0; i < ncls; i++)
+      weights[i] = cls_tv.data[i];
+    if (!classes_.IsDefined()) {
+      classes.resize(ncls);
+      int cls = 0;
+      for (int i = 0; i < ncls; i++, cls++) {
+        if (cls == background)
+          cls++;
+        classes[i] = cls;
+      }
+    }
+  }
+}
+
+template <int ndim>
+int RandomObjectBBox::PickBox(span<Box<ndim, int>> boxes, int sample_idx) {
+  auto beg = boxes.begin();
+  auto end = boxes.end();
+  if (threshold_.IsDefined()) {
+    vec<ndim, int> threshold;
+    const int *thresh = threshold_[sample_idx].data;
+    assert(threshold_.get().shape[sample_idx] == TensorShape<1>{ ndim });
+    for (int i = 0; i < ndim; i++)
+      threshold[i] = thresh[i];
+    end = std::remove_if(beg, end, [threshold](auto box) {
```

Comment:
Any difference if the `auto box` was a reference instead of pass by value?

Reply:
It probaly should be.